presto-main/src/main/java/com/facebook/presto/operator/LookupJoiner.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.io.Closeable;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.concurrent.MoreFutures.tryGetFutureValue;
+import static java.util.Objects.requireNonNull;
+
+public final class LookupJoiner
```

Comment:
What is the purpose of this refactoring? Extract join specific logic in a separate class? Essentially, LookupJoinOperator now contains logic around OperatorContext, type and close.

Logic remained in the LookupJoinOperator looks reusable and quite useful. Why not to have a base class containing this logic instead?

Seems like it would be nice to extract "finishing" logic as well, first two checks in addInput to LookupJoinOperator and leave LookupJoiner with only join related logic. With base class one can check for `finishing` directly, while in your model - may be passing it in `getOutput(finishing)`.

So I think LookupJoiner still contains some operator related logic that is worth moving.